# Desafío 2 - Properati

## Modelo base: Precios hedónicos

Con el objeto de interpretar las relaciones entre la variable objetivo y las descriptivas, partimos de un modelo relativamente sencillo, conocido como Modelo de Precios Hedónicos y lo ejecutamos con statsmodels para verificar la significancia de las variables. Dicho modelo tiene la forma:
$$ \ln{(\text{price_usd_per_m2})} = \beta_0+\beta_1\times\text{rooms}+\beta_2\times\text{surface_total_in_m2}+\beta_{3i}\times\text{property_type}_i+\beta_{4i}\times\text{localidad}_i$$

donde $\text{property_type}$ y $\text{localidad}$ son variables categóricas.

Para ello, importamos el dataset limpio y las librerías necesarias

In [1]:
import numpy as np
import pandas as pd
import geopandas as gpd
import seaborn as sns
import matplotlib.pyplot as plt
import statsmodels.formula.api as smf

In [5]:
df = pd.read_csv('df2.csv')
df

Unnamed: 0 property_type       place_name  geonames_id     price  \
0               0            PH        Mataderos    3430787.0   62000.0   
1               2     apartment        Mataderos    3430787.0   72000.0   
2               4     apartment           Centro    3435548.0   64000.0   
3               5         house     Gualeguaychú    3433657.0       NaN   
4               6            PH            Munro    3430511.0  130000.0   
...           ...           ...              ...          ...       ...   
98916      121215     apartment         Belgrano    3436077.0  870000.0   
98917      121216         house           Beccar    3436080.0  498000.0   
98918      121217     apartment    Villa Urquiza    3433775.0  131500.0   
98919      121218     apartment      Plaza Colón          NaN   95900.0   
98920      121219     apartment  Capital Federal    3433955.0  129000.0   

      currency  price_aprox_local_currency  price_aprox_usd  \
0          USD                  1093959.00          62000.0   
1          USD                  1270404.00          72000.0   
2          USD                  1129248.00          64000.0   
3          USD                         NaN              NaN   
4          USD                  2293785.00         130000.0   
...        ...                         ...              ...   
98916      USD                 15350715.00         870000.0   
98917      USD                  8786961.00         498000.0   
98918      USD                  2320251.75         131500.0   
98919      USD                  1692107.55          95900.0   
98920      USD                  2276140.50         129000.0   

       surface_total_in_m2  surface_covered_in_m2  ...  Estrenar  Gimnasio  \
0                     55.0                   40.0  ...         0         0   
1                     55.0                   55.0  ...         0         0   
2                     35.0                   35.0  ...         0         0   
3                     53.0                   53.0  ...         0         0   
4                    106.0                   78.0  ...         0         0   
...                    ...                    ...  ...       ...       ...   
98916                113.0                   93.0  ...         0         0   
98917                360.0                  360.0  ...         0         0   
98918                 46.0                   39.0  ...         0         0   
98919                 48.0                   48.0  ...         0         0   
98920                 77.0                   77.0  ...         0         0   

       Lavadero Parrilla Pileta  SUM  Seguridad  log_price  \
0             0        0      0    0          0  11.034890   
1             0        0      0    0          0  11.184421   
2             0        0      0    0          0  11.066638   
3             1        0      0    0          0        NaN   
4             0        0      0    0          0  11.775290   
...         ...      ...    ...  ...        ...        ...   
98916         0        1      1    0          1  13.676248   
98917         0        1      1    0          0  13.118355   
98918         1        1      0    0          0  11.786762   
98919         1        0      0    0          0  11.471061   
98920         0        0      0    0          0  11.767568   

      log_price_aprox_usd log_price_usd_per_m2  
0               11.034890             7.027556  
1               11.184421             7.177088  
2               11.066638             7.511290  
3                     NaN                  NaN  
4               11.775290             7.111851  
...                   ...                  ...  
98916           13.676248             8.948861  
98917           13.118355             7.232251  
98918           11.786762             7.958121  
98919           11.471061             7.599860  
98920           11.767568             7.423762  

[98921 rows x 34 columns]

In [6]:
for i in df.columns:
    print(i, df[i].dtype)

Unnamed: 0 int64
property_type object
place_name object
geonames_id float64
price float64
currency object
price_aprox_local_currency float64
price_aprox_usd float64
surface_total_in_m2 float64
surface_covered_in_m2 float64
price_usd_per_m2 float64
price_per_m2 float64
rooms float64
description object
title object
lat float64
lon float64
precio_regex float64
moneda object
provincia object
localidad object
barrio object
Amenities int64
Cochera int64
Estrenar int64
Gimnasio int64
Lavadero int64
Parrilla int64
Pileta int64
SUM int64
Seguridad int64
log_price float64
log_price_aprox_usd float64
log_price_usd_per_m2 float64


##### Nos aseguramos que la variable `price_usd_per_m2` esté completa y obtenemos las variables logarítmicas actualizadas
<p style="color:red;"><b> NO EJECUTAR </p>

In [4]:
#df['price_usd_per_m2'] = df.price_usd_per_m2.fillna(df.price/df.surface_total_in_m2)
#df['log_price'] = df.price.apply(np.log)
#df['log_price_aprox_usd'] = df.price_aprox_usd.apply(np.log)
#df['log_price_usd_per_m2'] = df.price_usd_per_m2.apply(np.log)

##### Generamos el modelo descrito arriba con statsmodels

In [7]:
model_base = smf.ols('log_price_usd_per_m2 ~ rooms + surface_total_in_m2 +surface_covered_in_m2 + C(property_type) + C(localidad)', data=df)
model_base.fit().summary2()

E:\Miniconda3\envs\ds37\lib\site-packages\statsmodels\base\model.py:1294: RuntimeWarning: invalid value encountered in true_divide
  return self.params / self.bse
E:\Miniconda3\envs\ds37\lib\site-packages\scipy\stats\_distn_infrastructure.py:901: RuntimeWarning: invalid value encountered in greater
  return (a < x) & (x < b)
E:\Miniconda3\envs\ds37\lib\site-packages\scipy\stats\_distn_infrastructure.py:901: RuntimeWarning: invalid value encountered in less
  return (a < x) & (x < b)
E:\Miniconda3\envs\ds37\lib\site-packages\scipy\stats\_distn_infrastructure.py:1892: RuntimeWarning: invalid value encountered in less_equal
  cond2 = cond0 & (x <= _a)


<class 'statsmodels.iolib.summary2.Summary'>
"""
                                  Results: Ordinary least squares
====================================================================================================
Model:                         OLS                           Adj. R-squared:              0.614     
Dependent Variable:            log_price_usd_per_m2          AIC:                         37371.5474
Date:                          2019-10-08 19:49              BIC:                         39676.9606
No. Observations:              27823                         Log-Likelihood:              -18406.   
Df Model:                      279                           F-statistic:                 159.9     
Df Residuals:                  27543                         Prob (F-statistic):          0.00      
R-squared:                     0.618                         Scale:                       0.22208   
----------------------------------------------------------------------------------------------------
                                                     Coef.  Std.Err.    t     P>|t|   [0.025  0.975]
----------------------------------------------------------------------------------------------------
Intercept                                            7.2585   0.0704 103.0774 0.0000  7.1205  7.3966
C(property_type)[T.apartment]                        0.4409   0.0114  38.7563 0.0000  0.4186  0.4632
C(property_type)[T.house]                           -0.3786   0.0130 -29.1444 0.0000 -0.4040 -0.3531
C(property_type)[T.store]                            0.2874   0.0579   4.9648 0.0000  0.1740  0.4009
C(localidad)[T.Achiras]                             -0.0000   0.0000  -1.2652 0.2058 -0.0000  0.0000
C(localidad)[T.Adolfo Alsina]                        0.0000   0.0000   8.0925 0.0000  0.0000  0.0000
C(localidad)[T.Agronomía]                            0.0521   0.1241   0.4199 0.6746 -0.1912  0.2954
C(localidad)[T.Agua Blanca]                          0.0000   0.0000  11.2982 0.0000  0.0000  0.0000
C(localidad)[T.Agua de Oro]                          0.0000   0.0000   2.4596 0.0139  0.0000  0.0000
C(localidad)[T.Aguas Verdes]                        -1.3964   0.2046  -6.8242 0.0000 -1.7975 -0.9953
C(localidad)[T.Aldea Brasilera]                     -1.2228   0.4764  -2.5666 0.0103 -2.1566 -0.2890
C(localidad)[T.Aldea Salto]                          0.0000   0.0000   1.9789 0.0478  0.0000  0.0000
C(localidad)[T.Alejandro Roca]                       3.2256   0.4765   6.7694 0.0000  2.2916  4.1596
C(localidad)[T.Almafuerte]                           0.0000   0.0000   2.0020 0.0453  0.0000  0.0000
C(localidad)[T.Almagro]                              0.0371   0.0724   0.5116 0.6089 -0.1049  0.1790
C(localidad)[T.Almirante Brown]                     -0.3615   0.0787  -4.5951 0.0000 -0.5157 -0.2073
C(localidad)[T.Alta Gracia]                         -1.8126   0.4764  -3.8045 0.0001 -2.7464 -0.8787
C(localidad)[T.Aluminé]                             -0.0000   0.0000  -2.8819 0.0040 -0.0000 -0.0000
C(localidad)[T.Alvear]                              -0.2394   0.4765  -0.5024 0.6154 -1.1734  0.6945
C(localidad)[T.Anisacate]                            0.0000   0.0000   1.8067 0.0708 -0.0000  0.0000
C(localidad)[T.Apóstoles]                           -0.3091   0.4765  -0.6488 0.5165 -1.2430  0.6248
C(localidad)[T.Arequito]                            -0.3497   0.4764  -0.7342 0.4628 -1.2834  0.5840
C(localidad)[T.Arocena]                              0.0000   0.0000   2.2048 0.0275  0.0000  0.0000
C(localidad)[T.Arroyo Ceibal]                       -0.0000   0.0000  -2.7180 0.0066 -0.0000 -0.0000
C(localidad)[T.Arroyo Leyes]                        -0.0000   0.0000  -0.7046 0.4811 -0.0000  0.0000
C(localidad)[T.Arroyo Seco]                         -1.2233   0.4765  -2.5676 0.0102 -2.1572 -0.2895
C(localidad)[T.Ascochinga]                          -0.0000   0.0000  -4.1488 0.0000 -0.0000 -0.0000
C(localidad)[T.Ataliva]                        

Como podemos observar en el output anterior, el modelo se está fiteando sobre $27823$ observaciones y estamos describiendo un $0.609 (R^2=0.612)$ de la variación en el precio de los inmuebles.

A continuación probamos ingresar las dummies de Amenities, Seguridad, Cochera, A Estrenar, Gimnasio, Lavadero, Parrilla, Pileta, Sum y una transformación sobre rooms para aumentar el poder descriptivo del modelo. También optamos por no calcular un intercepto, ya que no tiene mucho sentido.

In [10]:
model_base = smf.ols('log_price ~ rooms + surface_total_in_m2 +surface_covered_in_m2 + C(property_type) + C(localidad)', data=df)
model_base.fit().summary2()

<class 'statsmodels.iolib.summary2.Summary'>
"""
                                  Results: Ordinary least squares
====================================================================================================
Model:                         OLS                         Adj. R-squared:                0.401     
Dependent Variable:            log_price                   AIC:                           77464.0077
Date:                          2019-10-08 19:59            BIC:                           80579.4017
No. Observations:              41353                       Log-Likelihood:                -38371.   
Df Model:                      360                         F-statistic:                   77.89     
Df Residuals:                  40992                       Prob (F-statistic):            0.00      
R-squared:                     0.406                       Scale:                         0.37782   
----------------------------------------------------------------------------------------------------
                                                     Coef.  Std.Err.    t     P>|t|   [0.025  0.975]
----------------------------------------------------------------------------------------------------
Intercept                                           11.0396   0.0830 133.0063 0.0000 10.8769 11.2023
C(property_type)[T.apartment]                        0.1257   0.0122  10.3441 0.0000  0.1019  0.1496
C(property_type)[T.house]                            0.2969   0.0135  22.0125 0.0000  0.2704  0.3233
C(property_type)[T.store]                            0.4605   0.0670   6.8698 0.0000  0.3291  0.5918
C(localidad)[T.Achiras]                              0.3384   0.4425   0.7647 0.4444 -0.5289  1.2056
C(localidad)[T.Adolfo Alsina]                        0.0000   0.0000   0.1913 0.8483 -0.0000  0.0000
C(localidad)[T.Agronomía]                            0.1773   0.1349   1.3146 0.1887 -0.0871  0.4417
C(localidad)[T.Agua Blanca]                         -0.1994   0.6202  -0.3215 0.7478 -1.4150  1.0162
C(localidad)[T.Agua de Oro]                          0.0000   0.0000   1.3769 0.1685 -0.0000  0.0000
C(localidad)[T.Aguas Verdes]                        -0.5967   0.2465  -2.4212 0.0155 -1.0798 -0.1137
C(localidad)[T.Aldea Brasilera]                     -0.5274   0.6202  -0.8503 0.3951 -1.7430  0.6882
C(localidad)[T.Aldea Salto]                         -0.0000   0.0000  -1.1844 0.2363 -0.0000  0.0000
C(localidad)[T.Alejandro Roca]                       2.5075   0.6202   4.0428 0.0001  1.2918  3.7232
C(localidad)[T.Almafuerte]                          -2.2116   0.6203  -3.5653 0.0004 -3.4274 -0.9958
C(localidad)[T.Almagro]                              0.1559   0.0856   1.8224 0.0684 -0.0118  0.3236
C(localidad)[T.Almirante Brown]                     -0.0653   0.0918  -0.7119 0.4766 -0.2453  0.1146
C(localidad)[T.Alta Gracia]                         -0.0695   0.3183  -0.2185 0.8271 -0.6934  0.5543
C(localidad)[T.Aluminé]                             -0.0000   0.0000  -1.8778 0.0604 -0.0000  0.0000
C(localidad)[T.Alvear]                              -0.3291   0.6202  -0.5306 0.5957 -1.5448  0.8866
C(localidad)[T.Anisacate]                           -1.5421   0.4424  -3.4857 0.0005 -2.4093 -0.6750
C(localidad)[T.Apóstoles]                           -0.1174   0.6202  -0.1893 0.8498 -1.3331  1.0982
C(localidad)[T.Arequito]                             0.4946   0.6201   0.7975 0.4251 -0.7209  1.7101
C(localidad)[T.Arocena]                              0.0000   0.0000   2.0117 0.0443  0.0000  0.0000
C(localidad)[T.Arroyo Ceibal]                       -1.4359   0.6202  -2.3150 0.0206 -2.6516 -0.2202
C(localidad)[T.Arroyo Leyes]                         0.0000   0.0000   1.5495 0.1213 -0.0000  0.0000
C(localidad)[T.Arroyo Seco]                         -0.2202   0.6202  -0.3550 0.7226 -1.4358  0.9955
C(localidad)[T.Ascochinga]                          -0.7640   0.3645  -2.0962 0.0361 -1.4783 -0.0496
C(localidad)[T.Ataliva]                        

In [9]:
df['rooms2'] = df.rooms**2
df['rooms3'] = df.rooms**3
df['rooms4'] = df.rooms**4
model_base_mod =  smf.ols('log_price_usd_per_m2 ~ rooms + rooms2 + rooms3 + rooms4 + surface_total_in_m2 + \
                            surface_covered_in_m2 + C(property_type) + C(localidad) + Seguridad + \
                            Amenities + Cochera + Estrenar + Gimnasio + Lavadero + Parrilla + Pileta + SUM -1', data=df)
model_base_mod.fit().summary2()

<class 'statsmodels.iolib.summary2.Summary'>
"""
                                  Results: Ordinary least squares
====================================================================================================
Model:                         OLS                           Adj. R-squared:              0.627     
Dependent Variable:            log_price_usd_per_m2          AIC:                         36428.1117
Date:                          2019-10-08 19:52              BIC:                         38832.3282
No. Observations:              27823                         Log-Likelihood:              -17922.   
Df Model:                      291                           F-statistic:                 162.1     
Df Residuals:                  27531                         Prob (F-statistic):          0.00      
R-squared:                     0.631                         Scale:                       0.21459   
----------------------------------------------------------------------------------------------------
                                                     Coef.  Std.Err.    t     P>|t|   [0.025  0.975]
----------------------------------------------------------------------------------------------------
C(property_type)[PH]                                 7.3620   0.0719 102.3437 0.0000  7.2210  7.5030
C(property_type)[apartment]                          7.7575   0.0713 108.7501 0.0000  7.6177  7.8973
C(property_type)[house]                              6.9568   0.0719  96.8225 0.0000  6.8159  7.0976
C(property_type)[store]                              7.6121   0.0900  84.5434 0.0000  7.4356  7.7885
C(localidad)[T.Achiras]                              0.0000   0.0000   1.1678 0.2429 -0.0000  0.0000
C(localidad)[T.Adolfo Alsina]                       -0.0000   0.0000  -7.0432 0.0000 -0.0000 -0.0000
C(localidad)[T.Agronomía]                           -0.0324   0.1221  -0.2656 0.7905 -0.2718  0.2069
C(localidad)[T.Agua Blanca]                          0.0000   0.0000  11.3869 0.0000  0.0000  0.0000
C(localidad)[T.Agua de Oro]                         -0.0000   0.0000  -7.3928 0.0000 -0.0000 -0.0000
C(localidad)[T.Aguas Verdes]                        -1.4193   0.2013  -7.0518 0.0000 -1.8138 -1.0248
C(localidad)[T.Aldea Brasilera]                     -1.2957   0.4684  -2.7662 0.0057 -2.2138 -0.3776
C(localidad)[T.Aldea Salto]                          0.0000   0.0000   8.6526 0.0000  0.0000  0.0000
C(localidad)[T.Alejandro Roca]                       3.2122   0.4684   6.8578 0.0000  2.2941  4.1303
C(localidad)[T.Almafuerte]                          -0.0000   0.0000  -5.5863 0.0000 -0.0000 -0.0000
C(localidad)[T.Almagro]                              0.0512   0.0713   0.7186 0.4724 -0.0885  0.1909
C(localidad)[T.Almirante Brown]                     -0.3531   0.0774  -4.5625 0.0000 -0.5048 -0.2014
C(localidad)[T.Alta Gracia]                         -1.7786   0.4684  -3.7976 0.0001 -2.6966 -0.8606
C(localidad)[T.Aluminé]                              0.0000   0.0000   1.2834 0.1994 -0.0000  0.0000
C(localidad)[T.Alvear]                              -0.3784   0.4686  -0.8074 0.4194 -1.2969  0.5401
C(localidad)[T.Anisacate]                           -0.0000   0.0000  -3.9449 0.0001 -0.0000 -0.0000
C(localidad)[T.Apóstoles]                           -0.2989   0.4685  -0.6380 0.5235 -1.2172  0.6194
C(localidad)[T.Arequito]                            -0.4298   0.4684  -0.9176 0.3588 -1.3479  0.4883
C(localidad)[T.Arocena]                             -0.0000   0.0000  -4.2130 0.0000 -0.0000 -0.0000
C(localidad)[T.Arroyo Ceibal]                       -0.0000   0.0000  -0.8323 0.4052 -0.0000  0.0000
C(localidad)[T.Arroyo Leyes]                         0.0000   0.0000   3.5607 0.0004  0.0000  0.0000
C(localidad)[T.Arroyo Seco]                         -1.3735   0.4685  -2.9318 0.0034 -2.2917 -0.4553
C(localidad)[T.Ascochinga]                          -0.0000   0.0000  -1.6530 0.0983 -0.0000  0.0000
C(localidad)[T.Ataliva]                        

Revisamos por qué hay tantas localidades y barrios que no son significativos. Para ello agrupamos las observaciones por localidad, obtenemos los estadísitcos básicos, filtramos por las que no son NaN y las ordenamos por desvío estándar.

In [19]:
varianzas = df.groupby('localidad').log_price_usd_per_m2.describe()
display(varianzas.shape)
varianzas.loc[varianzas['std'].isnull()==False].sort_values('std', ascending=False)

(519, 8)

count      mean       std       min       25%  \
localidad                                                                 
Baradero                    6.0  6.898410  3.123209  3.717183  4.260327   
Zárate                     12.0  5.274652  2.851703  0.728528  3.029108   
La Cumbre                   2.0  5.478362  2.792535  3.503741  4.491052   
San Antonio de Arredondo   14.0  5.955985  2.485753  2.642299  4.925027   
Nono                        4.0  5.517404  2.309755  2.525729  4.361798   
...                         ...       ...       ...       ...       ...   
Villa Icho Cruz             2.0  5.401889  0.002530  5.400100  5.400995   
Dorrego                     3.0  5.636283  0.000000  5.636283  5.636283   
Manfredi                    2.0  5.184297  0.000000  5.184297  5.184297   
Timbúes                     3.0  4.448693  0.000000  4.448693  4.448693   
Pueblo Andino               2.0  3.824162  0.000000  3.824162  3.824162   

                               50%       75%        max  
localidad                                                
Baradero                  6.304836  9.761309  10.571317  
Zárate                    4.928395  7.787727   9.305651  
La Cumbre                 5.478362  6.465672   7.452982  
San Antonio de Arredondo  5.407662  6.732212  12.847927  
Nono                      5.876872  7.032478   7.790144  
...                            ...       ...        ...  
Villa Icho Cruz           5.401889  5.402783   5.403678  
Dorrego                   5.636283  5.636283   5.636283  
Manfredi                  5.184297  5.184297   5.184297  
Timbúes                   4.448693  4.448693   4.448693  
Pueblo Andino             3.824162  3.824162   3.824162  

[316 rows x 8 columns]

De las 519 localidades que tenemos, sólo tenemos datos del logaritmo de los precios en 270, lo cual explica el resultado. También podemos observar cuáles son las localidades con mayor desviación estándar.

In [10]:
model_caba_mod =  smf.ols('log_price_usd_per_m2 ~ C(rooms) + surface_total_in_m2 + \
                            surface_covered_in_m2 + C(property_type) + C(localidad) + Seguridad + \
                            Amenities + Cochera + Estrenar + Gimnasio + Lavadero + Parrilla + Pileta + SUM -1', data=df.loc[df.provincia=='Capital Federal'])
model_caba_mod.fit().summary2()

<class 'statsmodels.iolib.summary2.Summary'>
"""
                           Results: Ordinary least squares
=====================================================================================
Model:                   OLS                       Adj. R-squared:          0.542    
Dependent Variable:      log_price_usd_per_m2      AIC:                     4824.7299
Date:                    2019-10-07 21:19          BIC:                     5451.6631
No. Observations:        11798                     Log-Likelihood:          -2327.4  
Df Model:                84                        F-statistic:             167.4    
Df Residuals:            11713                     Prob (F-statistic):      0.00     
R-squared:               0.546                     Scale:                   0.087500 
-------------------------------------------------------------------------------------
                                      Coef.  Std.Err.    t     P>|t|   [0.025  0.975]
-------------------------------------------------------------------------------------
C(rooms)[1.0]                         7.4157   0.0451 164.4886 0.0000  7.3273  7.5041
C(rooms)[2.0]                         7.3615   0.0450 163.6095 0.0000  7.2733  7.4497
C(rooms)[3.0]                         7.3462   0.0450 163.3752 0.0000  7.2581  7.4344
C(rooms)[4.0]                         7.3068   0.0454 160.9153 0.0000  7.2178  7.3958
C(rooms)[5.0]                         7.2793   0.0469 155.2463 0.0000  7.1874  7.3713
C(rooms)[6.0]                         7.2593   0.0503 144.3765 0.0000  7.1608  7.3579
C(rooms)[7.0]                         7.1618   0.0546 131.2458 0.0000  7.0548  7.2687
C(rooms)[8.0]                         7.1646   0.0689 103.9297 0.0000  7.0295  7.2997
C(rooms)[9.0]                         7.1611   0.1158  61.8440 0.0000  6.9341  7.3881
C(rooms)[10.0]                        7.1051   0.1065  66.7054 0.0000  6.8963  7.3139
C(rooms)[11.0]                        6.5507   0.1423  46.0210 0.0000  6.2717  6.8297
C(rooms)[12.0]                        7.0594   0.1161  60.8285 0.0000  6.8319  7.2869
C(rooms)[13.0]                        6.5683   0.3096  21.2121 0.0000  5.9613  7.1752
C(rooms)[14.0]                        6.4475   0.2155  29.9142 0.0000  6.0250  6.8700
C(rooms)[15.0]                        7.0583   0.2123  33.2538 0.0000  6.6423  7.4744
C(rooms)[17.0]                       -0.0000   0.0000  -4.4324 0.0000 -0.0000 -0.0000
C(rooms)[18.0]                        0.0000   0.0000  23.5161 0.0000  0.0000  0.0000
C(rooms)[22.0]                       -0.0000   0.0000 -27.4541 0.0000 -0.0000 -0.0000
C(property_type)[T.apartment]         0.2254   0.0111  20.3040 0.0000  0.2037  0.2472
C(property_type)[T.house]            -0.1033   0.0181  -5.6961 0.0000 -0.1389 -0.0678
C(property_type)[T.store]             0.1098   0.0450   2.4384 0.0148  0.0215  0.1980
C(localidad)[T.Agronomía]            -0.0053   0.0782  -0.0677 0.9460 -0.1586  0.1480
C(localidad)[T.Almagro]               0.0772   0.0458   1.6878 0.0915 -0.0125  0.1669
C(localidad)[T.Balvanera]            -0.0460   0.0469  -0.9815 0.3264 -0.1378  0.0459
C(localidad)[T.Barracas]             -0.0727   0.0479  -1.5169 0.1293 -0.1665  0.0212
C(localidad)[T.Barrio Norte]          0.3340   0.0457   7.3151 0.0000  0.2445  0.4234
C(localidad)[T.Belgrano]              0.3512   0.0450   7.8036 0.0000  0.2630  0.4394
C(localidad)[T.Boca]                 -0.2559   0.0533  -4.7991 0.0000 -0.3604 -0.1514
C(localidad)[T.Boedo]                 0.3022   0.0505   5.9853 0.0000  0.2032  0.4011
C(localidad)[T.Caballito]             0.1303   0.0450   2.8931 0.0038  0.0420  0.2185
C(localidad)[T.Catalinas]            -0.0619   0.2994  -0.2068 0.8362 -0.6487  0.5249
C(localidad)[T.Centro / Microcentro]  0.0376   0.0509   0.7384 0.4603 -0.0622  0.1375
C(localidad)[T.Chacarita]             0.0635   0.0539   1.1776 0.2390 -0.0422  0.1691
C(localidad)[T.Coghlan]               0.2216   0.0526   4.2098 0.0000  0.1184  0.3248
C(localidad)[T.Colegia

In [20]:
df.loc[df.provincia=='Capital Federal'].groupby('localidad').log_price_usd_per_m2.describe().sort_values('std', ascending=False)

count      mean       std       min       25%  \
localidad                                                              
Villa Santa Rita        48.0  8.176110  1.540745  7.057396  7.487582   
Versalles               54.0  7.881811  1.207798  6.579251  7.282578   
Catalinas                3.0  8.364431  1.064820  7.513891  7.767324   
Monte Castro            72.0  7.666624  1.056871  2.713281  7.307741   
Pompeya                 55.0  7.319088  0.932873  6.127597  6.856832   
Villa Riachuelo          5.0  7.445846  0.911166  6.645391  7.114370   
Villa Lugano           178.0  7.188193  0.831523  5.907123  6.843883   
Paternal               157.0  7.696419  0.809844  6.406980  7.418581   
Velez Sarsfield         35.0  7.489080  0.772381  6.655531  7.041887   
Boedo                  262.0  7.781244  0.735100  6.781004  7.393263   
Villa Devoto           352.0  7.749789  0.678997  6.319969  7.500052   
Centro / Microcentro   207.0  7.709526  0.671074  3.687629  7.522627   
Parque Chacabuco       123.0  7.458546  0.645599  4.511750  7.234019   
Villa del Parque       408.0  7.741294  0.613086  4.510474  7.528519   
Parque Avellaneda       52.0  7.314871  0.587603  5.352385  7.077538   
Boca                   196.0  7.321095  0.558293  3.798694  7.218718   
Once                   156.0  7.580087  0.540142  5.033286  7.313220   
Congreso               250.0  7.527983  0.521569  2.916716  7.358947   
Mataderos              326.0  7.244834  0.520176  6.214608  6.900968   
San Cristobal          323.0  7.562278  0.504947  6.496459  7.373578   
Tribunales              84.0  7.673925  0.502869  6.907755  7.417262   
Las Cañitas            127.0  8.151006  0.491779  6.519097  7.949209   
Parque Patricios       260.0  7.525025  0.473162  4.807511  7.379526   
Floresta               409.0  7.430223  0.468346  2.805112  7.289123   
Monserrat              488.0  7.663854  0.461479  3.891820  7.536364   
Retiro                 222.0  7.945757  0.449845  6.343638  7.719005   
Flores                1055.0  7.575220  0.446873  3.057773  7.418581   
Liniers                268.0  7.440097  0.444428  3.004865  7.252027   
San Nicolás            162.0  7.678010  0.439515  5.710376  7.470103   
Villa Pueyrredón       141.0  7.593812  0.437108  4.795499  7.390607   
Agronomía               37.0  7.649371  0.428567  6.757309  7.545333   
Palermo               3132.0  8.085477  0.427656  3.443583  7.873943   
Villa Ortuzar           84.0  7.779250  0.423900  6.886921  7.604697   
Recoleta              1295.0  8.076154  0.419036  5.233770  7.871415   
Barracas               369.0  7.486212  0.417559  5.994979  7.313220   
Belgrano              2294.0  8.001411  0.410986  3.218876  7.846119   
Nuñez                  617.0  7.995338  0.410882  5.521461  7.777774   
Puerto Madero          521.0  8.686645  0.406596  6.502290  8.534191   
Barrio Norte           987.0  7.965096  0.405027  4.794892  7.760187   
Villa General Mitre     45.0  7.510994  0.403265  6.250976  7.408531   
Constitución           146.0  7.414765  0.401493  6.246107  7.148917   
San Telmo              935.0  7.706851  0.397714  3.136707  7.584382   
Villa Crespo          1025.0  7.817365  0.391749  4.483785  7.661527   
Caballito             1833.0  7.783787  0.389012  3.264546  7.646362   
Villa Urquiza         1136.0  7.826513  0.384249  5.137085  7.696607   
Coghlan                208.0  7.795228  0.379437  4.853522  7.697474   
Almagro                848.0  7.667475  0.379318  3.045523  7.547734   
Villa Luro             256.0  7.595186  0.379280  6.332391  7.409405   
Villa Soldati           14.0  6.791484  0.361602  5.999833  6.656445   
Balvanera              603.0  7.527348  0.357312  3.144728  7.385110   
Villa Real              37.0  7.329473  0.348708  6.294651  7.052937   
Colegiales             317.0  7.796177  0.343775  4.877177  7.675924   
Parque Chas             44.0  7.742108  0.320333  7.130899  7.636120   
Saavedra               446.0  7.752834  0.314968  6.761801  7.6

##### Generamos dummies por localidad (barrio), tipo de inmueble y cantidad de ambientes para incorporar en scikit learn
Construimos dos dataframes nuevos: 
- prices (que posee todas las variables relacionadas a precio)
- X (que posee todas las varialbes descriptivas a partir de las cuales obtener las dummies)

In [12]:
#df.drop(columns='Unnamed: 0', inplace=True)
prices = df.filter(['price', 'price_aprox_local_currency', 'price_aprox_usd','price_usd_per_m2','price_per_m2', \
               'log_price', 'log_price_aprox_usd', 'log_price_usd_per_m2', 'log_price_usd_per_m2', 'log_price_usd_per_m2 '])
#pd.get_dummies(df, columns=['property_type', 'localidad'], drop_first=True)
X = df.filter(['property_type', 'surface_total_in_m2', 'surface_covered_in_m2', 'rooms','localidad',\
               'barrio', 'Amenities', 'Cochera', 'Estrenar', 'Gimnasio','Lavadero', 'Parrilla', 'Pileta', 'SUM', \
               'Seguridad', 'rooms2', 'rooms3', 'rooms4' ])
X = pd.get_dummies(X.property_type, drop_first=1).join(X).drop('property_type', axis=1)
X = pd.get_dummies(X.localidad, drop_first=1).join(X).drop('localidad', axis=1)

In [13]:
from sklearn.model_selection import train_test_split, KFold, cross_val_score
from sklearn.linear_model import LinearRegression, Lasso, LassoCV, Ridge, RidgeCV
from sklearn.preprocessing import PolynomialFeatures, StandardScaler
from sklearn.metrics import mean_squared_error, r2_score


def findAlphaLasso(X,y,randomState=53,tries=5,alphaFrom=0.00001,alphaTo=100000,steps=1000):
    kf = KFold(n_splits=5, shuffle=True, random_state=randomState)
    step_value=(alphaTo-alphaFrom)/steps
    print("Lasso:")
    prevAlphaTest=0
    for i in range(0,tries):
        al_lasso = np.linspace(alphaFrom, alphaTo, steps)
        lm_lasso_cv= LassoCV(alphas=al_lasso, cv=kf, normalize=True)
        lm_lasso_cv.fit(X, y)
        if( round(prevAlphaTest*10000) == round(lm_lasso_cv.alpha_*10000) ):
            return lm_lasso_cv 
        prevAlphaTest=lm_lasso_cv.alpha_
        alphaFrom = prevAlphaTest - step_value*2
        alphaTo = prevAlphaTest   + step_value*2
        step_value=(alphaTo-alphaFrom)/steps
        print("intento {} value {} from {} to {} ".format(i+1,lm_lasso_cv.alpha_,alphaFrom,alphaTo))
    return lm_lasso_cv

In [14]:
X_train, X_test, y_train, y_test = train_test_split(X, prices.log_price_usd_per_m2, shuffle=True)
lasso = findAlphaLasso(X_train, y_train)
print(" Score Train Lasso: %.2f\n" % lasso.score(X_train, y_train))
print(" Score Test Lasso: %.2f\n" % lasso.score(X_test, y_test))

Lasso:


ValueError: Input contains NaN, infinity or a value too large for dtype('float64').